packages/storage/src/providers/s3/utils/validateMultipartUploadXML.ts

```diff
@@ -0,0 +1,56 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { IntegrityError } from '../../../errors/IntegrityError';
+
+import { parser } from './client/runtime';
+import { CompletedMultipartUpload } from './client/s3data/types';
+import {
+	deserializeCompletedPartList,
+	emptyArrayGuard,
+	map,
+} from './client/utils';
+
+export function validateMultipartUploadXML(
+	input: CompletedMultipartUpload,
+	xml: string,
+) {
+	const parsedXML = parser.parse(xml);
+	if (!parsedXML.Part) {
+		throw new IntegrityError();
+	}
+	const mappedParts: any[] = map(parsedXML, {
+		Part: [
+			'Part',
+			value => emptyArrayGuard(value, deserializeCompletedPartList),
+		],
+	}).Part;
+	if (mappedParts.length !== input.Parts?.length) {
+		throw new IntegrityError();
+	}
+	if (input.Parts?.length) {
+		const mappedPartsMap = mappedParts.reduce((acc, part) => {
+			acc[part.PartNumber] = part;
+
+			return acc;
+		}, {});
+		for (const p of input.Parts) {
+			const mappedPart = mappedPartsMap[p.PartNumber || ''];
+			if (p.ETag !== mappedPart?.ETag) {
+				throw new IntegrityError();
+			}
+			if (p.ChecksumCRC32 !== mappedPart?.ChecksumCRC32) {
+				throw new IntegrityError();
+			}
+			if (p.ChecksumCRC32C !== mappedPart?.ChecksumCRC32C) {
+				throw new IntegrityError();
+			}
+			if (p.ChecksumSHA1 !== mappedPart?.ChecksumSHA1) {
+				throw new IntegrityError();
+			}
+			if (p.ChecksumSHA256 !== mappedPart?.ChecksumSHA256) {
+				throw new IntegrityError();
+			}
```

Comment:
If we're going to try to be completely thorough, we should create a deep equals helpers.

Reply:
replaced with `isEqual` from lodash